README.md
# lookup-ssh-pubkey

## Background

If you are using ssh you probably already know of public key
authentication. For this to work you normally have a file
`$HOME/.ssh/authorized_keys` in each users homedirectory.

While this works pretty well it can be hard to ship these
public keys. This gets even more hard when user home
directories are not guranteed to even exist.

Imagine you have 20 LDAP users. The homedirectory will only
be created after the first login. So you cannot deploy the
`authorized_keys` file beforehand (or your deployment
mechanism needs to create all homedirectories first).

For this scenario you can now deploy all keys to a common
directory and use a custom command to retrieve the keys.

## How to use

The `lookup-ssh-pubkey` executable can be used as a local
lookup method to retrieve public authorized ssh keys.

It is intended to be used by openssh. In `sshd_config` just
specify

    AuthorizedKeysCommand /path/to/lookup-ssh-pubkey
    AuthorizedKeysCommandUser nobody

The command will check if there is a file

    /etc/ssh-public-keys.d/<username>.pub

present that should have the same format as a
normale `.ssh/authorized_keys` file.

## Security considerations

The public keys should be **readable** for everyone as the
`lookup-ssh-pubkey` should normally run as user nobody.
The keyfiles itself should only be writeable by the owner and
should be owned by either root or the corresponding user the
keyfile belongs to. This is important so no user can write
its own public key to the `authorized_keys` file of a different
user.

The `lookup-ssh-pubkey` will raise an error if this condition is
not met. The same goes for each parent directory. If a parent
directory would be writeable by someone else, an attacker would
be able to remove a key (or a part of the directory tree) and
replace it with a new keyfile (or directory tree). This
condition will also be tested.

## How to build

you can build and install the package locally with the following
commands

    mkdir build
    cd build
    cmake ..
    make
    make install



debian/changelog
lookup-ssh-pubkey (1.0.0) UNRELEASED; urgency=low

  * inital version

 -- Stefan Schulte <stschulte@posteo.de>  Thu, 25 May 2017 20:16:54 +0100


src/authorized_keys.c
#include <errno.h>
#include <stdlib.h>
#include <stdio.h>

#include <libgen.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <limits.h>
#include <pwd.h>

#include "config.h"

#define BUFSIZE 2048

/*
 * Check if a given keyfile is secure. A keyfile is considered unsecure if
 * it is not owned by the user or by root. It is also unsecure if the
 * keyfile is writeable by group or other.
 *
 * The same check will be done for every parent directory. If a
 * parentdirectory would be writeable by an attacker, this user would be
 * able to replace the keyfile or the whole directory tree with a custom
 * version.
 *
 * Returns 0 on success and -1 on failure
 */
int is_keyfile_secure(int fd, uid_t uid, const char *filename, char *err, size_t errlen) {
  struct stat st;
  struct passwd* owner;
  char* parentdir;
  char path[PATH_MAX];

  if(realpath(filename, path) == NULL) {
    snprintf(err, errlen, "realpath %s failed: %s", filename, strerror(errno));
    return -1;
  }

  if(fstat(fd, &st) < 0) {
    snprintf(err, errlen, "cannot stat file %s: %s", path, strerror(errno));
    return -1;
  }

  if(!S_ISREG(st.st_mode)) {
    snprintf(err, errlen, "%s is not a regular file", path);
    return -1;
  }

  if(st.st_uid != 0 && st.st_uid != uid) {
    if((owner = getpwuid(st.st_uid)) != NULL)
      snprintf(err, errlen, "%s is owned by unexpected user %s", path, owner->pw_name);
    else
      snprintf(err, errlen, "%s is owned by unknown user %u", path, st.st_uid);

    return -1;
  }
  if((st.st_mode & 022) != 0) {
    snprintf(err, errlen, "%s should only be writeable by the owner itself (mode=%03o)", path, st.st_mode & 07777);
    return -1;
  }

  /* now check all leading directories until we reach the root directory */
  for(;;) {
    if((parentdir = dirname(path)) == NULL) {
      snprintf(err, errlen, "dirname(%s) failed", filename);
      return -1;
    }

    /* make sure we set path to the new parent directory so it will use
     * that as the base for the next interation */
    strncpy(path, parentdir, sizeof(path)-1);
    path[sizeof path - 1] = '\0';

    if(stat(parentdir, &st) < 0) {
      snprintf(err, errlen, "unable to stat directory %s", parentdir);
      return -1;
    }
    else if(st.st_uid !=0 && st.st_uid != uid) {
      snprintf(err, errlen, "bad ownership for directory %s", parentdir);
      return -1;
    }
    else if((st.st_mode & 022) != 0) {
      snprintf(err, errlen, "bad mode (0%3o) for directory %s", st.st_mode & 07777, parentdir);
      return -1;
    }

    if((strcmp(parentdir, "/") == 0) || (strcmp(parentdir, ".") == 0))
      break;
  }
  return 0;
}

void cat(FILE* f) {
  char buffer[BUFSIZE];
  size_t rbytes;

  while((rbytes = fread(buffer, sizeof(char), BUFSIZE, f)) != 0) {
    fwrite(buffer, sizeof(char), rbytes, stdout);
  }
}

void usage(void) {
  fprintf(stderr,
    "usage: "PROGRAM" (-h |  --help)\n"
    "usage: "PROGRAM" --version\n"
    "usage: "PROGRAM" <user>\n"
    "\n"
    "retrieve the public keys of the specified user\n"
    "\n"
    "the intention of this script is to be used as an AuthorizedKeysCommand\n"
    "by the ssh daemon. It will retrieve the public key file of the specified\n"
    "user from "KEYDIR". If no key can be found it will return with a zero\n"
    "exitcode. If a key can be found it will be printed to stdout\n");
}

int main(int argc, char** argv) {
  char* username = NULL;

  struct passwd* user;
  char keyfile[PATH_MAX];
  char errbuf[1024];
  int i = 1;
  FILE* f;

  while(i < argc) {
    if(*argv[i] != '-')
      break;

    if(strcmp(argv[i], "--") == 0) {
      i++;
      break;
    }

    if(strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
      usage();
      exit(EXIT_SUCCESS);
    }
    else if(strcmp(argv[i], "--version") == 0) {
      printf(PROGRAM " version " VERSION "\n");
      exit(EXIT_SUCCESS);
    }
    else {
      fprintf(stderr, PROGRAM ": unknown argument: %s\n", argv[i]);
      exit(EXIT_FAILURE);
    }
  }

  if(argc < 2) {
    usage();
    exit(EXIT_SUCCESS);
  }
  else if(i == argc - 1) {
    username = argv[i];
  }
  else {
    fprintf(stderr, PROGRAM ": wrong number of arguments\n");
    exit(EXIT_FAILURE);
  }


  user = getpwnam(username);
  if(user == NULL) {
    fprintf(stderr, "authorized_keys: user %s could not be found\n", username);
    return 1;
  }

  snprintf(keyfile, PATH_MAX, "%s/%s.pub", KEYDIR, user->pw_name);

  if((f = fopen(keyfile, "r")) == NULL) {
    /* if there is no file or we cannot open it, make sure we do not fail and
     * openssh can continue with the normal authorized_keys behaviour */
    return 0;
  }

  if(is_keyfile_secure(fileno(f), user->pw_uid, keyfile, errbuf, sizeof(errbuf)) == 0) {
    cat(f);
    fclose(f);
  }
  else {
    fprintf(stderr, PROGRAM" : %s\n", errbuf);
    fclose(f);
    return -1;
  }

  return 0;
}


src/lookup-ssh-pubkey.c
#include <errno.h>
#include <stdlib.h>
#include <stdio.h>

#include <libgen.h>
#include <string.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <fcntl.h>
#include <unistd.h>
#include <limits.h>
#include <pwd.h>

#include "config.h"

#define BUFSIZE 2048

/*
 * Check if a given keyfile is secure.
 *
 * A keyfile is considered secure, if only root (or the user the key belongs
 * to) can write to file. It is considerd unsecure if any other user or group
 * can write to it (reading the keyfile is no problem, since we are talking
 * public keys here).
 *
 * The same check will be done for every parent directory. If a
 * parentdirectory would be writeable by an attacker, this user would be
 * able to replace the keyfile or the whole directory tree with a custom
 * version.
 *
 * Returns 0 on success and -1 on failure
 */
int is_keyfile_secure(int fd, const char *filename, uid_t expected_uid, char *err, size_t errlen) {
  struct stat st;
  struct passwd* owner;
  char* parentdir;
  char path[PATH_MAX];

  if(realpath(filename, path) == NULL) {
    snprintf(err, errlen, "realpath %s failed: %s", filename, strerror(errno));
    return -1;
  }

  if(fstat(fd, &st) < 0) {
    snprintf(err, errlen, "cannot stat file %s: %s", path, strerror(errno));
    return -1;
  }

  if(!S_ISREG(st.st_mode)) {
    snprintf(err, errlen, "%s is not a regular file", path);
    return -1;
  }

  if(st.st_uid != 0 && st.st_uid != expected_uid) {
    if((owner = getpwuid(st.st_uid)) != NULL)
      snprintf(err, errlen, "%s is owned by unexpected user %s", path, owner->pw_name);
    else
      snprintf(err, errlen, "%s is owned by unknown user %u", path, st.st_uid);

    return -1;
  }
  if((st.st_mode & 022) != 0) {
    snprintf(err, errlen, "%s should only be writeable by the owner itself (mode=%03o)", path, st.st_mode & 07777);
    return -1;
  }

  /* now check all leading directories until we reach the root directory */
  for(;;) {
    if((parentdir = dirname(path)) == NULL) {
      snprintf(err, errlen, "dirname(%s) failed", filename);
      return -1;
    }

    /* make sure we set path to the new parent directory so it will use
     * that as the base for the next interation */
    strncpy(path, parentdir, sizeof(path)-1);
    path[sizeof path - 1] = '\0';

    if(stat(parentdir, &st) < 0) {
      snprintf(err, errlen, "unable to stat directory %s", parentdir);
      return -1;
    }
    else if(st.st_uid !=0 && st.st_uid != expected_uid) {
      snprintf(err, errlen, "bad ownership for directory %s", parentdir);
      return -1;
    }
    else if((st.st_mode & 022) != 0) {
      snprintf(err, errlen, "bad mode (0%3o) for directory %s", st.st_mode & 07777, parentdir);
      return -1;
    }

    if((strcmp(parentdir, "/") == 0) || (strcmp(parentdir, ".") == 0))
      break;
  }
  return 0;
}

void cat(FILE* f) {
  char buffer[BUFSIZE];
  size_t rbytes;

  while((rbytes = fread(buffer, sizeof(char), BUFSIZE, f)) != 0) {
    fwrite(buffer, sizeof(char), rbytes, stdout);
  }
}

void usage(void) {
  fprintf(stderr,
    "usage: "PROGRAM" (-h |  --help)\n"
    "usage: "PROGRAM" --version\n"
    "usage: "PROGRAM" <user>\n"
    "\n"
    "retrieve the public keys of the specified user\n"
    "\n"
    "the intention of this script is to be used as an AuthorizedKeysCommand\n"
    "by the ssh daemon. It will retrieve the public key file of the specified\n"
    "user from "KEYDIR". If no key can be found it will return with a zero\n"
    "exitcode. If a key can be found it will be printed to stdout\n");
}

int main(int argc, char** argv) {
  char* username = NULL;   /* the username we want to retrieve the keys for */

  struct passwd* user;     /* the user info if present */
  char keyfile[PATH_MAX];  /* path to the potential keyfile */
  FILE* f;                 /* file handle for the keyfile */

  char errbuf[1024];
  int i = 1;

  while(i < argc) {
    if(*argv[i] != '-')
      break;

    if(strcmp(argv[i], "--") == 0) {
      i++;
      break;
    }

    if(strcmp(argv[i], "-h") == 0 || strcmp(argv[i], "--help") == 0) {
      usage();
      exit(EXIT_SUCCESS);
    }
    else if(strcmp(argv[i], "--version") == 0) {
      printf(PROGRAM " version " VERSION "\n");
      exit(EXIT_SUCCESS);
    }
    else {
      fprintf(stderr, PROGRAM ": unknown argument: %s\n", argv[i]);
      exit(EXIT_FAILURE);
    }
  }

  if(argc < 2) {
    usage();
    exit(EXIT_SUCCESS);
  }
  else if(i == argc - 1) {
    username = argv[i];
  }
  else {
    fprintf(stderr, PROGRAM ": wrong number of arguments\n");
    exit(EXIT_FAILURE);
  }


  user = getpwnam(username);
  if(user == NULL) {
    fprintf(stderr, PROGRAM ": user %s could not be found\n", username);
    return EXIT_FAILURE;
  }

  snprintf(keyfile, PATH_MAX, "%s/%s.pub", KEYDIR, user->pw_name);

  if((f = fopen(keyfile, "r")) == NULL) {
    /* if there is no file or we cannot open it, make sure we do not fail and
     * openssh can continue with the normal authorized_keys behaviour */
    return EXIT_SUCCESS;
  }

  if(is_keyfile_secure(fileno(f), keyfile, user->pw_uid, errbuf, sizeof(errbuf)) == 0) {
    cat(f);
    fclose(f);
  }
  else {
    fprintf(stderr, PROGRAM ": %s\n", errbuf);
    fclose(f);
    return EXIT_FAILURE;
  }

  return EXIT_SUCCESS;
}
